subprojects/normalization-java/src/main/java/org/gradle/api/internal/changedetection/state/LineEndingAwareResourceHasher.java

```diff
@@ -0,0 +1,209 @@
+        hasher.putString(getClass().getName());
+        hasher.putString(LineEndingNormalizingInputStream.class.getName());
+        hasher.putString(lineEndingSensitivity.name());
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(RegularFileSnapshotContext snapshotContext) {
+        return lineEndingSensitivity.isCandidate(snapshotContext.getSnapshot()) ?
+            hashContent(snapshotContext) :
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    @Override
+    public HashCode hash(ZipEntryContext zipEntryContext) throws IOException {
+        return lineEndingSensitivity.isCandidate(zipEntryContext) ?
+            hashContent(zipEntryContext) :
+            delegate.hash(zipEntryContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(RegularFileSnapshotContext snapshotContext) {
+        NormalizedContentInfo normalizedContentInfo = collectContentInfo(new File(snapshotContext.getSnapshot().getAbsolutePath()));
+        return normalizedContentInfo.getContentType() == FileContentType.TEXT ?
+            normalizedContentInfo.getHash() :
+            delegate.hash(snapshotContext);
+    }
+
+    @Nullable
+    private HashCode hashContent(ZipEntryContext zipEntryContext) throws IOException {
+        NormalizedContentInfo normalizedContentInfo = zipEntryContext.getEntry().withInputStream(this::collectContentInfo);
+        return normalizedContentInfo.getContentType() == FileContentType.TEXT ?
+            normalizedContentInfo.getHash() :
+            delegate.hash(zipEntryContext);
+    }
+
+
+    public NormalizedContentInfo collectContentInfo(InputStream inputStream) {
+        FileContentTypeDetectingInputStream contentTypeDetectingInputStream = new FileContentTypeDetectingInputStream(inputStream);
+        ShortCircuitingInputStream<FileContentTypeDetectingInputStream> shortCircuitingInputStream = new ShortCircuitingInputStream<>(
+            contentTypeDetectingInputStream,
+            stream -> stream.getContentType() == FileContentType.BINARY
+        );
+        InputStream normalizedInputStream = new LineEndingNormalizingInputStream(shortCircuitingInputStream);
+
+        try {
+            HashCode hashCode = streamHasher.hash(normalizedInputStream);
+            return new NormalizedContentInfo(hashCode, contentTypeDetectingInputStream.getContentType(), shortCircuitingInputStream.isShortCircuited());
+        } finally {
+            closeAndIgnore(contentTypeDetectingInputStream, shortCircuitingInputStream, normalizedInputStream);
+        }
+    }
+
+    public NormalizedContentInfo collectContentInfo(File file) {
+        FileInputStream inputStream = null;
+        try {
+            inputStream = new FileInputStream(file);
+            return collectContentInfo(inputStream);
+        } catch (FileNotFoundException e) {
+            throw new UncheckedIOException(String.format("Failed to create MD5 hash for file '%s' as it does not exist.", file), e);
```

Comment:
Why would we want to handle this here specifically, as opposed to just relying the default behavior? Is the only reason to re-wrap the checked exception?

Reply:
Because of the different handling.  FileNotFound is a fundamentally different error than simply failing to close the file (which we were previously ignoring).  It's a moot point now, though, since these methods now just throw IOException and we catch it further up.